server/container/core/src/main/java/org/apache/james/server/core/BufferedDeferredFileOutputStream.java

```diff
@@ -0,0 +1,264 @@
+  * <p>
+  * This copy is done because 1. the original class is not written in the way that
+  * we can decorate file output stream. 2. when switched to FileOutputStream the
+  * original class does not buffer file output stream which could cause up to 50%
+  * performance loss compare to buffered stream (see JIRA JAMES-2343 for simple
+  * benchmark).
+  * <p>
+  * The only difference is in the {@link #thresholdReached()} where when updating
+  * {@link #currentOutputStream}, instead of directly assign FileOutputStream to
+  * it, here we wrap the file output stream with BufferedOutputStream.
+  *
+  * @link https://issues.apache.org/jira/browse/JAMES-2343
+ */
+public class BufferedDeferredFileOutputStream extends ThresholdingOutputStream {
+
+    /**
+     * The output stream to which data will be written prior to the theshold
+     * being reached.
+     */
+    private ByteArrayOutputStream memoryOutputStream;
+
+
+    /**
+     * The output stream to which data will be written at any given time. This
+     * will always be one of <code>memoryOutputStream</code> or
+     * <code>diskOutputStream</code>.
+     */
+    private OutputStream currentOutputStream;
+
+
+    /**
+     * The file to which output will be directed if the threshold is exceeded.
+     */
+    private File outputFile;
+
+    /**
+     * The temporary file prefix.
+     */
+    private final String prefix;
+
+    /**
+     * The temporary file suffix.
+     */
+    private final String suffix;
+
+    /**
+     * The directory to use for temporary files.
+     */
+    private final File directory;
+
+
+    /**
+     * True when close() has been called successfully.
+     */
+    private boolean closed = false;
+
+    /**
+     * Constructs an instance of this class which will trigger an event at the
+     * specified threshold, and save data to a file beyond that point.
+     *
+     * @param threshold  The number of bytes at which to trigger an event.
+     * @param outputFile The file to which data is saved beyond the threshold.
+     */
+    public BufferedDeferredFileOutputStream(final int threshold, final File outputFile) {
+        this(threshold,  outputFile, null, null, null);
+    }
+
+
+    /**
+     * Constructs an instance of this class which will trigger an event at the
+     * specified threshold, and save data to a temporary file beyond that point.
+     *
+     * @param threshold  The number of bytes at which to trigger an event.
+     * @param prefix Prefix to use for the temporary file.
+     * @param suffix Suffix to use for the temporary file.
+     * @param directory Temporary file directory.
+     */
+    public BufferedDeferredFileOutputStream(final int threshold, final String prefix, final String suffix, final File directory) {
+        this(threshold, null, prefix, suffix, directory);
+        if (prefix == null) {
+            throw new IllegalArgumentException("Temporary file prefix is missing");
+        }
```

Comment:
Preconditions

Reply:
The BufferedDefferedFileOutputStream is (mostly) taken from commons-io, I would like to get that code as close as possible from the original version, if you don't mind.